src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportOrchestratorJob.cs

```diff
@@ -0,0 +1,160 @@
+        {
+            EnsureArg.IsNotNull(queueClient, nameof(queueClient));
+            EnsureArg.IsNotNull(searchService, nameof(searchService));
+            EnsureArg.IsNotNull(loggerFactory, nameof(loggerFactory));
+
+            _queueClient = queueClient;
+            _searchService = searchService;
+            _logger = loggerFactory.CreateLogger<ExportOrchestratorJob>();
+        }
+
+        internal int NumberOfSurrogateIdRanges { get; set; } = DefaultNumberOfSurrogateIdRanges;
+
+        public async Task<string> ExecuteAsync(JobInfo jobInfo, IProgress<string> progress, CancellationToken cancellationToken)
+        {
+            var record = JsonConvert.DeserializeObject<ExportJobRecord>(jobInfo.Definition);
+            record.QueuedTime = jobInfo.CreateDate; // get record of truth
+            var surrogateIdRangeSize = (int)record.MaximumNumberOfResourcesPerQuery;
+            var groupJobs = await _queueClient.GetJobByGroupIdAsync((byte)QueueType.Export, jobInfo.GroupId, true, cancellationToken);
+
+            // for parallel case we enqueue in batches, so we should handle not completed registration
+            if (record.ExportType == ExportJobType.All && record.IsParallel && (record.Filters == null || record.Filters.Count == 0))
+            {
+                var atLeastOneWorkerJobRegistered = false;
+
+                var resourceTypes = string.IsNullOrEmpty(record.ResourceType)
+                                  ? (await _searchService.GetUsedResourceTypes(cancellationToken)).Select(_ => _.Name)
+                                  : record.ResourceType.Split(',');
+                resourceTypes = resourceTypes.OrderByDescending(_ => string.Equals(_, "Observation", StringComparison.OrdinalIgnoreCase)); // true first, so observation is processed as soon as
+
+                var since = record.Since == null ? new PartialDateTime(DateTime.MinValue).ToDateTimeOffset() : record.Since.ToDateTimeOffset();
+                var globalStartId = _searchService.GetSurrogateId(since.DateTime);
+                var till = record.Till.ToDateTimeOffset();
+                var globalEndId = _searchService.GetSurrogateId(till.DateTime) - 1; // -1 is so _till value can be used as _since in the next time based export
+
+                var enqueued = groupJobs.Where(_ => _.Id != jobInfo.Id) // exclude coord
+                                        .Select(_ => JsonConvert.DeserializeObject<ExportJobRecord>(_.Definition))
+                                        .Where(_ => _.EndSurrogateId != null) // This is to handle current mock tests. It is not needed but does not hurt.
+                                        .GroupBy(_ => _.ResourceType)
+                                        .ToDictionary(_ => _.Key, _ => _.Max(r => long.Parse(r.EndSurrogateId)));
+
+                await Parallel.ForEachAsync(resourceTypes, new ParallelOptions { MaxDegreeOfParallelism = 4, CancellationToken = cancellationToken }, async (type, cancel) =>
```

Comment:
Is there a reason why MaxDegreeOfParallelism is hardcoded to 4? Would this be something that we want to have configurable in the future?

Reply:
Unlike number of parallel tasks this is not a strong parameter. I do not expect any differences if value is increased.  So as long as it does not crash for low end systems at 4, we should never have need to change it.